pytorch_lightning/trainer/supporters.py

```diff
@@ -177,3 +182,315 @@ def to_disk(self) -> None:
             # Write predictions for current file to disk
             with fs.open(filepath, "wb") as fp:
                 torch.save(outputs, fp)
+
+
+class CycleIterator(object):
+    """
+    Iterator for restarting a dataloader if it runs out of samples
+    """
+    def __init__(self, loader: Any, length: int = None):
+        """
+
+        Args:
+            loader: the loader to restart for cyclic (and optionally infinite) sampling
+            length: the number of batches to sample (with restarted loaders if necessary) before raising StopIteration
+                if None: infinite
+
+        """
+        if length is None:
+            length = float('inf')
+
+        self.length = length
+        self.loader = loader
+        self._loader_iter = None
+        self.counter = 0
+
+    def __iter__(self) -> Any:
+        """
+
+        Creates the internal iterator and returns self
+
+        Returns:
+            CycleIterator: self
+
+        """
+        self.counter = 0
+        self._loader_iter = iter(self.loader)
+        return self
+
+    def __next__(self) -> Any:
+        """
+        Fetches the next batch from internal dataloader and restarts
+        it if necessary
+
+        Returns:
+            Any: the resulting batch
+
+        Raises:
+            StopIteration: if more then :attr:`length` batches have been returned
+
+        """
+        # Note: if self.length is `inf`, then the iterator will never stop
+        if self.counter >= self.__len__():
+            raise StopIteration
+
+        try:
+            return next(self._loader_iter)
+
+        except StopIteration:
+            self._loader_iter = iter(self.loader)
+            return next(self._loader_iter)
+
+        finally:
+            self.counter += 1
+
+    def __len__(self) -> int:
```

Comment:
```suggestion
    def __len__(self) -> Union[int, float]:
```
as eaflier you define `length = float('inf')`